internal/controller/supervisor.go

```diff
@@ -0,0 +1,140 @@
+package controller
+
+import (
+	"context"
+	"time"
+
+	"github.com/hashicorp/consul/lib/retry"
+)
+
+// flapThreshold is the minimum amount of time between restarts for us *not* to
+// consider a controller to be stuck in a crash-loop.
+const flapThreshold = 2 * time.Second
+
+// supervisor keeps a task running, restarting it on-error, for as long as the
+// given lease is held. When the lease is lost, the context given to the task
+// will be canceled. If the task persistently fails (i.e. the controller is in
+// a crash-loop) supervisor will use exponential backoff to delay restarts.
+type supervisor struct {
+	task  task
+	lease Lease
+
+	running    bool
+	startedAt  time.Time
+	errCh      chan error
+	cancelTask context.CancelFunc
+
+	backoff        *retry.Waiter
+	backoffUntil   time.Time
+	backoffTimerCh <-chan time.Time
+}
```

Comment:
I'm kind of curious from a design standpoint why you have a single supervisor manage the lifecycle of a single controller. 

I can't say that I'm too familiar with the overhead or the ability to reason about more goroutines (vs less), but just some stats that have gotten me thinking about the complexity and overhead:

- 1 goroutine per supervisor
- 1 goroutine per controllerRunner
- 2 goroutines per controller implementation (guessing here)

Let's say we have 20 resource types in the system that quickly becomes 80 goroutines. 

Just to consider having less moving pieces, is it possible to have a single supervisor manage the lifecycles of all the controllers since in it's current state, raft leadership applies uniformly to all of the controllers or is this a future looking separation or concern?

Reply:
Great question! There are a couple of interesting topics here 🤓

Generally speaking, goroutines are cheap and can be used much more liberally than real OS threads. They have mininal memory overhead ([2KB for the stack](https://github.com/golang/go/blob/8f2db14cd35bbd674cb2988a508306de6655e425/src/runtime/stack.go#L72)) and Go's scheduler is very efficient. The HTTP stack even spins up goroutines [for each connection](https://cs.opensource.google/go/go/+/refs/tags/go1.20.3:src/net/http/server.go;l=3089), so it's not uncommon for production systems to have tens or even hundreds of thousands of running goroutines per process.

That said, you can take it too far, and we've certainly seen scheduler-related latency and resource starvation before.

I tend not to worry about spinning up a handful of goroutines in cases like this, because it's fixed number (you cannot add new controllers at runtime) and a much lower order of magnitude than, say, the number of connections we handle.

Even without separate leases, I think it would add quite a bit of complexity to track multiple controllers per-supervisor,  which is how I settled on this design.

Now, onto leases...

I think the current model of running every controller on the Raft leader is going to become infeasible quite quickly as teams migrate their features to the controller architecture. So I don't think it'll be long before we need to support a method of leader election that's more granular than piggybacking on Raft's leader election, which is where the lease abstraction comes in.

Hope that all makes sense! Shout if you'd like to chat more synchronously.